src/Lecture2.hs

```diff
@@ -264,7 +353,21 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval vars expr = case expr of
+    Lit x -> Right x
+    Var sx ->
+        let mx = lookup sx vars
+        in case mx of
+            Just x -> Right x
+            Nothing -> Left (VariableNotFound sx)
+    Add l r ->
+        let lRes = eval vars l
+            rRes = eval vars r
+        in case (lRes, rRes) of
+            (Left err, _) -> Left err
+            (_, Left err) -> Left err
+            (Right lx, Right rx) -> Right (lx + rx)
```

Comment:
That's a great helper function! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️